pkg/controller/che/tls-secrets.go

```diff
@@ -0,0 +1,287 @@
+			logrus.Errorf("Failed to set owner for \"%s\" secret. Error: %s", cheTLSSecretName, err)
+			return reconcile.Result{}, err
+		}
+		if err := r.client.Update(context.TODO(), cheTLSSecret); err != nil {
+			logrus.Errorf("Failed to update owner for \"%s\" secret. Error: %s", cheTLSSecretName, err)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// ===== Check Che CA certificate ===== //
+
+	cheTLSSelfSignedCertificateSecret := &corev1.Secret{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: checluster.Namespace, Name: cheTLSSelfSignedCertificateSecretName}, cheTLSSelfSignedCertificateSecret)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			// Error reading Che self-signed secret info
+			logrus.Errorf("Error getting Che self-signedsecert \"%s\": %v", cheTLSSecretName, err)
+			return reconcile.Result{}, err
+		}
+		// Che CA self-signed cetificate secret doesn't exist.
+		// Such situation could happen between reconcile loops, when CA cert is deleted.
+		// However the certificates should be created together,
+		// so it is mandatory to remove Che TLS secret now and recreate the pair.
+		cheTLSSecret = &corev1.Secret{}
+		err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: checluster.Namespace, Name: cheTLSSecretName}, cheTLSSecret)
+		if err != nil { // No need to check for not found error as the secret already exists at this point
+			logrus.Errorf("Error getting Che secert \"%s\": %v", cheTLSSecretName, err)
+			return reconcile.Result{}, err
+		}
+		if err = r.client.Delete(context.TODO(), cheTLSSecret); err != nil {
+			logrus.Errorf("Error deleting Che TLS secret \"%s\": %v", cheTLSSecretName, err)
+			return reconcile.Result{}, err
+		}
+		// Invalid secrets cleaned up, recreate them now
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// Che CA self-signed certificate secret exists, check for required data fields
+	if !isCheCASecretValid(cheTLSSelfSignedCertificateSecret) {
+		logrus.Infof("Che self signed certificate secret \"%s\" is invalid. Recrating...", cheTLSSelfSignedCertificateSecretName)
+		// Che CA self-signed certificate secret is invalid, delete it
+		if err = r.client.Delete(context.TODO(), cheTLSSelfSignedCertificateSecret); err != nil {
+			logrus.Errorf("Error deleting Che self-signed secret \"%s\": %v", cheTLSSelfSignedCertificateSecretName, err)
+			return reconcile.Result{}, err
+		}
+		// Also delete Che TLS as the certificates should be created together
+		// Here it is not mandatory to check Che TLS secret existence as it is handled above
+		if err = r.client.Delete(context.TODO(), cheTLSSecret); err != nil {
+			logrus.Errorf("Error deleting Che TLS secret \"%s\": %v", cheTLSSecretName, err)
+			return reconcile.Result{}, err
+		}
+		// Regenerate Che TLS certicates and recreate secrets
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// Check owner reference
+	if cheTLSSelfSignedCertificateSecret.ObjectMeta.OwnerReferences == nil {
```

Comment:
You set up owner references twice. Logic is very similar, maybe make sense to create separated method?

Reply:
Yes, the two blocks have identical logic except: secret configuration object, secret name and error messages. Also we have to deal with returning values in case of error so in places where we would call such method we need to add new check for an error. Taking into account that this two pieces of code are small and we would still have to handle errors (it is not possible to turn calls into one line) I would leave them as is. But if you have strong opposite vision, I may rework this part, just let me know.